### PR TITLE
Add local notification support

### PR DESCRIPTION
**Pull request by satya164**
**TODO**

- [x] Implement `scheduleNotification` on Android
- [x] implement `cancelAllNotifications` on Android
- [ ] Add `remote: false` property on Android
- [ ] Add `data` property on Android

cc @grabbou

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,7 +24,7 @@ Please use Node 6 and npm 3. We recommend installing Node using [nvm](https://gi
 
 - `npm install` in the `js` and `tools-public` directories.
 - Install [the Gulp CLI](http://gulpjs.com/) globally: `npm i gulp-cli -g`.
-- Run the packager with `cd tools-public && gulp`. Leave this running while you run the clients. The clients access this packager using your computer's IP, so make sure that your computer and device are on the same WiFi network.
+- Run the packager with `cd tools-public && npm run gulp`. Leave this running while you run the clients. The clients access this packager using your computer's IP, so make sure that your computer and device are on the same WiFi network.
```
Comment (jesseruder) [inline README.md]: Don't want to change this, this will break things internally

File: android/exponentview/build.gradle [hunk 2]
```diff
@@ -167,7 +167,7 @@ android {
 
 task generateDynamicMacros(type: Exec) {
   // This gets run from the 'app' directory
-  commandLine 'bash', '../../tools-public/generate-dynamic-macros-android.sh'
+  commandLine '../../tools-public/generate-dynamic-macros-android.sh'
```
Comment (jesseruder) [inline android/exponentview/build.gradle]: We'll need to make sure this works internally for us

File: android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/NotificationsModule.java [hunk 3]
```diff
@@ -73,8 +96,142 @@ public void onEventFailure(String errorMessage) {
...
+      promise.reject(NO_ACTIVITY_ERROR, "Activity not found");
+      return;
+    }
+
+    Intent intent;
+
+    if (details.hasKey("link")) {
+      intent = new Intent(Intent.ACTION_VIEW, Uri.parse(details.getString("link")));
+    } else {
+      Class activityClass = currentActivity.getClass();
```
Comment (jesseruder) [inline android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/NotificationsModule.java]: don't think this is right, should be LauncherActivity

File: android/run.sh [hunk 4]
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
 
-./gradlew ${1:-installDevDebug} && adb shell am start -n host.exp.exponent/host.exp.exponent.LauncherActivity
+./gradlew ${1:-installDevDebug} --stacktrace && adb shell am start -n host.exp.exponent/host.exp.exponent.LauncherActivity
```
Comment (jesseruder) [inline android/run.sh]: what does this do?

File: tools-public/package.json [hunk 5]
```diff
@@ -18,6 +19,7 @@
     "babel-preset-es2016-node5": "^1.1.2",
     "co": "^4.6.0",
     "gulp": "gulpjs/gulp#4.0",
+    "gulp-cli": "^1.2.2",
```
Comment (jesseruder) [inline tools-public/package.json]: you should just install gulp-cli globally, we don't want to add this

File: tools-public/package.json [hunk 6]
```diff
@@ -4,6 +4,7 @@
   "private": true,
   "description": "",
   "scripts": {
+    "gulp": "gulp",
```
Comment (jesseruder) [inline tools-public/package.json]: remove

File: tools-public/generate-dynamic-macros-android.sh [hunk 7]
```diff
@@ -8,5 +8,6 @@ fi
 
 pushd ../../tools-public/
 mkdir -p ../android/exponentview/src/main/java/host/exp/exponent/generated/
-gulp generate-dynamic-macros --buildConstantsPath ../android/exponentview/src/main/java/host/exp/exponent/generated/ExponentBuildConstants.java --platform android
+
+npm run gulp -- generate-dynamic-macros --buildConstantsPath ../android/exponentview/src/main/java/host/exp/exponent/generated/ExponentBuildConstants.java --platform android
```
Comment (jesseruder) [inline tools-public/generate-dynamic-macros-android.sh]: don't want to change this

File: android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/NotificationsModule.java [hunk 8]
```diff
@@ -73,8 +96,142 @@ public void onEventFailure(String errorMessage) {
...
+      String manifestUrl = (String) mExperienceProperties.get(KernelConstants.MANIFEST_URL_KEY);
+      intent = new Intent(getReactApplicationContext(), activityClass);
+      intent.putExtra(KernelConstants.MANIFEST_URL_KEY, manifestUrl);
+    }
+
+    PendingIntent contentIntent = PendingIntent.getActivity(context, 0, intent, 0);
+    builder.setContentIntent(contentIntent);
+
+
+    JSONObject notificationPreferences = mManifest.optJSONObject(ExponentManifest.MANIFEST_NOTIFICATION_INFO_KEY);
```
Comment (jesseruder) [inline android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/NotificationsModule.java]: we should separate this (color, icon) out into a shared utility that `ExponentGcmListenerService` also uses

**Review by jesseruder — CHANGES_REQUESTED** [review]
A few changes but overall looks good. Thanks!

File: android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/NotificationsModule.java (hunk 3 above)
Comment (satya164) [inline android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/NotificationsModule.java]: @jesseruder this is code inside exponent SDK which won't have access to `LauncherActivity` which is in the app.

File: android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/NotificationsModule.java (hunk 8 above)
Comment (satya164) [inline android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/NotificationsModule.java]: @jesseruder will do

File: android/run.sh (hunk 4 above)
Comment (satya164) [inline android/run.sh]: @jesseruder this prints more details when building fails

**Comment by satya164** [comment]
Ok, will revert my gulp changes :D

File: android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/NotificationsModule.java (hunk 3 above)
Comment (jesseruder) [inline android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/NotificationsModule.java]: Hmm ok. In that case can we do `Class activityClass = Class.forName(KernelConstants.MAIN_ACTIVITY_NAME)`

and then in KernelConstants:
`public static String MAIN_ACTIVITY_NAME = "host.exp.exponentview.ExponentActivity";`

and then in ExponentApplication:
`KernelConstants.MAIN_ACTIVITY_NAME = "host.exp.exponent.LauncherActivity";`

File: android/run.sh (hunk 4 above)
Comment (jesseruder) [inline android/run.sh]: sounds good

**Comment by grabbou** [comment]
@jesseruder added iOS part to the PR, thought you may want to double-check and submit some requests for changes.

**Comment by satya164** [comment]
I'll take care of the review comments soon.

**Comment by jesseruder** [comment]
@terribleben is going to look at the ios part

File: ios/Exponent/ExponentView/ExponentViewManager.m [hunk 9]
```diff
@@ -115,6 +116,12 @@ - (void)application:(UIApplication *)application didReceiveRemoteNotification:(N
   [[EXRemoteNotificationManager sharedInstance] handleRemoteNotification:notification fromBackground:isFromBackground];
```
Comment (terribleben) [inline ios/Exponent/ExponentView/ExponentViewManager.m]: Should probably also handle `UIApplicationLaunchOptionsLocalNotificationKey`

File: ios/Exponent/Versioned/Modules/Api/EXNotifications.m [hunk 10]
```diff
@@ -45,4 +63,74 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId
...
+  
+  [RCTSharedApplication() scheduleLocalNotification:notification];
+  
+  resolve(notification.userInfo[@"id"]);
+}
+
+RCT_EXPORT_METHOD(cancelNotification:(NSString *)uniqueId)
+{
+  for (UILocalNotification *notification in [RCTSharedApplication() scheduledLocalNotifications])
+  {
```
Comment (terribleben) [inline ios/Exponent/Versioned/Modules/Api/EXNotifications.m]: Style - don't put these brackets on their own line

File: ios/Exponent/Versioned/Modules/Api/Permissions/EXLocalNotificationRequester.m [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    @"expires": EXPermissionExpiresNever,
+  };
+}
+
+- (void)requestPermissionsWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(__unused RCTPromiseRejectBlock)reject
+{
+  UIUserNotificationType types = UIUserNotificationTypeBadge | UIUserNotificationTypeSound | UIUserNotificationTypeAlert;
+  [RCTSharedApplication() registerUserNotificationSettings:[UIUserNotificationSettings settingsForTypes:types categories:nil]];
+  
+  resolve([[self class] permissions]);
```
Comment (terribleben) [inline ios/Exponent/Versioned/Modules/Api/Permissions/EXLocalNotificationRequester.m]: Can't resolve here - the system may have shown a dialog to the user and we're waiting for them to decide whether to grant permissions. You need to listen to the callback methods on UIApplicationDelegate, then resolve/reject after taht. See `EXRemoteNotificationRequester` for an example of this pattern.

File: ios/Exponent/Environment/EXLocalNotificationManager.m [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+- (void)handleLocalNotification:(UILocalNotification *)notification fromBackground:(BOOL)isFromBackground
+{
+  NSDictionary *payload = notification.userInfo;
+  
+  if (payload) {
+    NSDictionary *body = [payload objectForKey:@"body"];
+    NSString *experienceId = [payload objectForKey:@"experienceId"];
+    if (body && experienceId) {
+      [[EXKernel sharedInstance] sendNotification:body toExperienceWithId:experienceId fromBackground:isFromBackground];
```
Comment (terribleben) [inline ios/Exponent/Environment/EXLocalNotificationManager.m]: Did we want to let the experience know this was a local notification as opposed to a remote one?

File: ios/Exponent/Versioned/Modules/Api/Permissions/EXLocalNotificationRequester.m [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+@end
+
+@implementation EXLocalNotificationRequester
+
++ (NSDictionary *)permissions
+{
+  UIUserNotificationSettings *currentSettings = RCTSharedApplication().currentUserNotificationSettings;
+  
+  EXPermissionStatus status = currentSettings != UIUserNotificationTypeNone
+    ? EXPermissionStatusGranted
```
Comment (terribleben) [inline ios/Exponent/Versioned/Modules/Api/Permissions/EXLocalNotificationRequester.m]: style - `?` and `:` go on the previous line

**Review by terribleben — CHANGES_REQUESTED** [review]

File: ios/Exponent/Environment/EXLocalNotificationManager.m (hunk 12 above)
Comment (grabbou) [inline ios/Exponent/Environment/EXLocalNotificationManager.m]: Yeah, that would be good. I think we could add `remote: false` property to the `Notification` itself. Easiest to support.

File: ios/Exponent/Versioned/Modules/Api/Permissions/EXLocalNotificationRequester.m (hunk 11 above)
Comment (grabbou) [inline ios/Exponent/Versioned/Modules/Api/Permissions/EXLocalNotificationRequester.m]: Valid point, thanks!

File: ios/Exponent/Environment/EXLocalNotificationManager.m (hunk 12 above)
Comment (terribleben) [inline ios/Exponent/Environment/EXLocalNotificationManager.m]: `remote: false` seems fine to me. @jesseruder or @brentvatne any opinion? Also seems easy to change later if we decide we want something different.

Comment (jesseruder) [inline ios/Exponent/Environment/EXLocalNotificationManager.m]: seems fine to me

File: android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsHelper.java [hunk 14]
```diff
@@ -0,0 +1,198 @@
+package versioned.host.exp.exponent.modules.api.notifications;
```
Comment (jesseruder) [inline android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsHelper.java]: can you move this out of the `versioned` directory? just put it under exponentview/host.exp.exponent

File: android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsManager.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
+package versioned.host.exp.exponent.modules.api.notifications;
```
Comment (jesseruder) [inline android/exponentview/src/main/java/versioned/host/exp/exponent/modules/api/notifications/NotificationsManager.java]: same thing here, move out of `versioned`

**Review by jesseruder — CHANGES_REQUESTED** [review]
A few comments on the Android side. Once these comments are fixed and you revert the gulp changes the Android side looks good @satya164. Thanks!

**Comment by grabbou** [comment]
Regarding my stuff - I am to implement it today (29th). Had a looong flight over the weekend, was going back to life :)

File: ios/Exponent/ExponentView/ExponentViewManager.m (hunk 9 above)
Comment (grabbou) [inline ios/Exponent/ExponentView/ExponentViewManager.m]: Done. Although I am not sure what's the `if (remoteNotification || badgeNumber > 0)` check responsible for (the 2nd part of it). Seems that in case of `nil || 1` it will pass null remote notification to the client. Is that desirable?

File: exponent-view-template/ios/exponent-view-template/AppDelegate.m [hunk 16]
```diff
@@ -55,4 +55,9 @@ - (void)application:(UIApplication *)application didReceiveRemoteNotification:(N
     [[ExponentViewManager sharedInstance] application:application didReceiveRemoteNotification:notification];
 }
 
+- (void)application:(UIApplication *)application didRegisterUserNotificationSettings:(UIUserNotificationSettings *)notificationSettings
+{
+    [[ExponentViewManager sharedInstance] application:application didRegisterUserNotificationSettings:notificationSettings];
+}
+
```
Comment (grabbou) [inline exponent-view-template/ios/exponent-view-template/AppDelegate.m]: @terribleben mind explaining a bit about the purpose of view-template?

**Comment by grabbou** [comment]
@satya164 one important change is that `Notification` object now has remote: boolean property. I think you should handle that as well. See the sdk PR.

@terribleben Added changes, still need to test it, but waiting to finish the review process first :)

File: android/exponentview/src/main/java/host/exp/exponent/notifications/NotificationsHelper.java [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+        case "day":
+          interval = DateUtils.DAY_IN_MILLIS;
+          break;
+        case "week":
+          interval = DateUtils.WEEK_IN_MILLIS;
+          break;
+        case "month":
+          interval = DateUtils.DAY_IN_MILLIS * 30;
+          break;
+        case "year":
```
Comment (satya164) [inline android/exponentview/src/main/java/host/exp/exponent/notifications/NotificationsHelper.java]: Who will need this!

Comment (jesseruder) [inline android/exponentview/src/main/java/host/exp/exponent/notifications/NotificationsHelper.java]: lol

**Review by jesseruder — COMMENTED** [review]
Android side looks good @satya164! Will wait for @terribleben to approve iOS side.

File: ios/Exponent/ExponentView/ExponentViewManager.m (hunk 9 above)
Comment (terribleben) [inline ios/Exponent/ExponentView/ExponentViewManager.m]: There is a null check inside `handleRemoteNotification:` so it won't pass anything to the experience. I think the `badgeNumber > 0` case is something I wrote in case we wanted to manually clear the badge number, but we don't currently use it. I will think about the best solution now that we have two handlers here.

File: exponent-view-template/ios/exponent-view-template/AppDelegate.m (hunk 16 above)
Comment (terribleben) [inline exponent-view-template/ios/exponent-view-template/AppDelegate.m]: This is a work in progress so don't worry about breaking/testing it right now. We are adding the ability to semi-`eject` to native code, so `ExponentViewManager` is an API that can work with any `AppDelegate`, not just Exponent's.

**Review by terribleben — APPROVED** [review]
Looks great, thanks a lot!

edit - also good thoughts about `remote` vs. `local` -- we're not sure which one we want yet, so let's just go ahead and merge (if @jesseruder approves) and we'll fix it later if we change our minds.

**Comment by grabbou** [comment]
Great, thanks @terribleben for the comments. I'll go ahead and test few edge cases I have in mind and if all goes well - I'll report back saying it's ready for merge!

**Comment by terribleben** [comment]
We went ahead and merged this - I think @jesseruder is going to follow up with that last bit. Thanks guys!
